internal/app/go-filecoin/internal/submodule/storage_protocol_submodule.go

```diff
@@ -51,40 +87,45 @@ func NewStorageProtocolSubmodule(
 	repoPath string,
 	sealProofType abi.RegisteredProof,
 	stateViewer *appstate.Viewer,
-) (*StorageProtocolSubmodule, error) {
+) error {
+	sm.pieceManager = pm
+
 	pnode := storagemarketconnector.NewStorageProviderNodeConnector(minerAddr, c.State, m.Outbox, mw, pm, s, stateViewer)
-	cnode := storagemarketconnector.NewStorageClientNodeConnector(cborutil.NewIpldStore(bs), c.State, mw, s, m.Outbox, clientAddr, stateViewer)
 
 	pieceStagingPath, err := paths.PieceStagingDir(repoPath)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// ensure pieces directory exists
 	err = os.MkdirAll(pieceStagingPath, 0700)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	fs, err := filestore.NewLocalFileStore(filestore.OsPath(pieceStagingPath))
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	dt := graphsyncimpl.NewGraphSyncDataTransfer(h, gsync)
-
-	provider, err := impl.NewProvider(smnetwork.NewFromLibp2pHost(h), ds, bs, fs, piecestore.NewPieceStore(ds), dt, pnode, minerAddr, sealProofType)
+	err = dt.RegisterVoucherType(reflect.TypeOf(&smvalid.StorageDataTransferVoucher{}), smvalid.NewProviderRequestValidator(statestore.New(ds)))
 	if err != nil {
-		return nil, errors.Wrap(err, "error creating graphsync provider")
+		return err
 	}
 
-	client, err := impl.NewClient(smnetwork.NewFromLibp2pHost(h), bs, dt, nil, nil, cnode)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating storage client")
-	}
+	sm.StorageProvider, err = impl.NewProvider(smnetwork.NewFromLibp2pHost(h), ds, bs, fs, piecestore.NewPieceStore(ds), dt, pnode, minerAddr, sealProofType)
+	return err
+}
 
-	return &StorageProtocolSubmodule{
-		StorageClient:   client,
-		StorageProvider: provider,
-	}, nil
+func (sm *StorageProtocolSubmodule) Provider() iface.StorageProvider {
+	return sm.StorageProvider
```

Comment:
Since this may be nil, suggest returning an error or ok boolean to alert the caller.

Reply:
Now returns an error if uninitialized.